Improved console

RunDevelopment opened the pull request:
# Improved console
I create an improved console modelled after the [Console Living Standard](https://console.spec.whatwg.org/).

## New JS functions
info, warn, debug, dir, clear, count, time, timeEnd, group, groupCollapsed, groupEnd

## Format specifiers
console now supports [Format specifiers](https://console.spec.whatwg.org/#formatting-specifiers) (This does not include %c). Further information can be found [here](https://developer.mozilla.org/de/docs/Web/API/Console#Using_string_substitutions) and [here](https://developers.google.com/web/tools/chrome-devtools/console/console-write#string_substitution_and_formatting).
Examples:
```js
var a = [1, 4.5, "abc", [1, 2], Array];

for (var i=0; i<a.length; i++)
    console.log("%d: %s", i, a[i]);
// 0: 1
// 1: 4.5
// 2: abc
// 3: 1,2
// 4: function Array() { [native code] }

for (var i=0; i<a.length; i++)
    console.log("%d: %d", i, a[i]);
// 0: 1
// 1: 4
// 2: 0
// 3: 0
// 4: 0

for (var i=0; i<a.length; i++)
    console.log("%d: %f", i, a[i]);
// 0: 1
// 1: 4.5
// 2: NaN
// 3: NaN
// 4: NaN

for (var i=0; i<a.length; i++)
    console.log("%d: %o", i, a[i]);
// 0: 1
// 1: 4.5
// 2: "abc"
// 3: Array (2) [ 1, 2 ]
// 4: function Array()

console.log("%d", Math.PI); // 3
console.log("%f", Math.PI); // 3.141592653589793
console.log("%.0f", Math.PI); // 3
console.log("%.2f", Math.PI); // 3.14
```

As a consequence function like log no longer simple output `[JSValue].ToString()`. Objects and Arrays will now display content and JS properties (using `Tools.JSValueToObjectString`), instead of displaying `[object Object]`.
Examples:
```js
var a = [1, "foo", new RegExp("[a-z]", "i")];
var obj = {bar: "bar", a: a};
console.log(a); // Array (3) [ 1, "foo", /[a-z]/i ]
console.log(obj); // Object { bar: "bar", a: Array[3] }
obj.obj = Object;
console.log("obj = ", obj); // obj = Object { bar: "bar", a: Array[3], obj: Object() }
```
Selfreferencial objects (and objects with object properties) are handled by only allowing a certain recursion depth before switching to a simplefied version that does not display content and properties. The default value for this maximal recursion depth is 1.
```js
var o = {};
o.o = o;
console.log(o); // Object { o: Object }
```

## Internal workings
count, time and group each need to store variables under a certain label. This makes it impractical to use static variables because different scripts could influence each other.
Because of this console is no longer a static class. Each instance of console is now tied to one GlobalContext (`CreateConsole `should be used over `new console`). `GlobalContext.ResetContext()` will automatically add a new console like before.
An instance of console will hold not only the variables for count, time and group but also the TextWriters to which the console will write, so different consoles can output to different TextWriters. (The default TextWriters will be `Console.Out` and `Console.Error`)

However, a console object is not directly passed to its GlobalContext. Instead a JSObject will be created that has all console functions (like log, error, etc.) but no other functions (like LogMessage).

RunDevelopment committed:
Added FormatArgs & JSValueToObjectString

RunDevelopment committed:
New console

RunDevelopment committed:
Updated to new console

nilproject commented:
If console is public from now then let's rename "console" (lower case) to "Console" (upper case for first letter) with regards to MS naming guidelines.
If you want to hide some functions from JS, you can use attribute HiddenAttribute. It's not necessory to create object and filter functions directly.
You can get current global context while any JS code is evaluating just by get value of `Context.CurrentContext.GlobalContext`. It's not necessory to store it in the object. FYI, in this engine only functions stores contexts. All other objects should be able to migrate from one context to others without side effects.

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,245 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+                    BaseLibrary.Function f = v.Value as BaseLibrary.Function;
+                    if (recursionDepth >= maxRecursionDepth)
+                        return f.name + "()";
+                    return $"function {f.name}()";
```
See Function.ToString(bool headerOnly)

RunDevelopment replied:
This is done. We now use Function.ToString()

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,245 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+                        typeName = typeName.Substring(8, typeName.Length - 1 - 8); // 8 = "[object ".Length
+
+                        if (recursionDepth >= maxRecursionDepth)
+                            return typeName;
```
JSValue.toString(...) not for all objects returns `[object TYPENAME]`. Maybe it would be useful to create function "GetTypeName" by extracting some code from JSValue.toString(...). With internal modifier, I think it can be modified in future.

RunDevelopment replied:
Well, it was hacky, to begin with.

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,245 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+
+                    if (include)
+                        s.Append(f.Substring(next, len));
+                    f = f.Substring(next + len);
```
We really need to create new line on every iteration?

RunDevelopment replied:
New line?
But, we really don't need to create a new string in every iteration.

nilproject replied:
Oops, my mistake. I meant string. But you understood me.

nilproject commented on NiL.JS/BaseLibrary/Console.cs:
```diff
@@ -1,68 +1,395 @@
+
+            return c;
+        }
+        public static console CreateConsole(GlobalContext globalContext, TextWriter logTw, TextWriter infoTw, TextWriter warnTw, TextWriter errorTw)
```
Dnt wrt bbrvtd nms fr pblc mmbrs.
Users often do not understand what this means.
Also it is not need to write type name in variable name.

Sometimes other programmers too.

nilproject commented on NiL.JS/BaseLibrary/Console.cs:
```diff
@@ -1,68 +1,395 @@
-        public static JSValue error(Arguments args)
+        
+
+        internal static List<console> _consoles;
```
We really need to store all instances in static field? I think, this solution can bring to memory leak.

RunDevelopment replied:
This is not relevant anymore.

nilproject commented:
I think, instead of allowing to override TextWritter, it will be useful to make all methods of Console virtual and allow users to do what they want.

nilproject commented on NiL.JS/BaseLibrary/Console.cs:
```diff
@@ -1,68 +1,395 @@
+            return JSValue.undefined;
         }
         
-        public static void asserta(Function f, JSValue sample)
```
asserta is used in some tests.

RunDevelopment committed:
Changed FunctionDefinition.ToString(true)

FunctionDefinition.ToString(true) will no longer return a trailing space.

RunDevelopment committed:
JSValueToObjectString now uses Function.ToString

RunDevelopment committed:
%O has now a greater recursion depth

This also changes how functions are printed.

RunDevelopment committed:
Added a Format Specifiers formatting test

RunDevelopment committed:
FormatArgs won't change the format string

RunDevelopment committed:
Added JSValueToString for %s

This function will first test if the value is an ObjectWrapper and will then use the native .ToString() method. Elsewise it will use the standard JSValue.ToString().

RunDevelopment committed:
Added GetTypeName() for JSValueToObjectString()

RunDevelopment committed:
Readded asserta

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1889,6 +1889,85 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+                case JSValueType.Function:
+                    {
+                        (v as Function).ToString();
+                        break;
```
Something went wrong.

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1889,6 +1889,85 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+            string typeName = v.toString(null).ToString(); // "[object TYPENAME]"
+            typeName = typeName.Substring(8, typeName.Length - 1 - 8); // 8 = "[object ".Length
+            
+                return typeName;
```
Why you extract typename from string, if you got it in code before?
This code will not be evaluated - all cases processed by `switch`.

RunDevelopment replied:
I forgot to clean it up, my bad.

RunDevelopment committed:
Cleaned up GetTypeName()

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,340 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+        }
+
+
+        public static string FormatArgs(Arguments args)
```
Let's do it more useful. Replace type of args with `IEnumerable` and extract first argument as `string`. If you make it public, then it should have a user-friendly interface. Create `Arguments` for every call of this function is pain.

RunDevelopment replied:
Agreed.

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,340 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+            return FormatArgs(args, start, stop, true);
+        }
+
+        public static string FormatArgs(Arguments args, int start, int stop, bool formatted)
```
I not found any real use this variant of function. I think `start`, `stop` and `formatted` are very specific parameters.

nilproject commented on NiL.JS/BaseLibrary/Console.cs:
```diff
@@ -1,44 +1,181 @@
+        public JSValue assert(JSValue thisBind, Arguments args)
+        {
+            if (!(bool)args[0])
+                LogArguments(LogLevel.Log, args, 1);
```
Just use `System.Linq`: `IEnumerable.Skip(1)`

RunDevelopment commented:
Maybe we could add a property Context.Console which will get the console instance from its GlobalContext which has the actual instance of console.
This would eliminate the need for all GlobalContext references in console.

RunDevelopment commented:
On the name:
I' against a capital "Console" because it was a pain to see every Console.WriteLine breaking down.
Maybe the simplest solution would be to rename it to JSConsole.
Maybe we could also move it to NiL.JS.Core, because it isn't really dependent on any types in BaseLibrary, plus if we do add a Context.Console property JSConsole will be closer to Context. It simply feels better for me.

nilproject commented:
We cannot add Context.Console and cannot move Console to NiL.JS.Core because console not allowed on .Net Portable. This platform is still supported and must have an identical Api.
Let's rename it to JSConsole.

RunDevelopment commented:
If .Net Portable doesn't support it, then shouldn't it be
```c#
#if !(PORTABLE || NETCORE)
```
instead of
```c#
#if !NETCORE
```
like in GlobalContext?

nilproject commented:
It's not needed. This file not included into project for .NET Portable.

RunDevelopment committed:
Updated FormatArgs

RunDevelopment committed:
Updated and renamed console to JSConsole

RunDevelopment committed:
Updated to JSConsole

nilproject commented:
Why you create `ConsoleObject`? Why not pass instance of `JSConsole` directly into JS environment? 
[See this example](https://github.com/nilproject/NiL.JS/blob/version-2.4/Examples/5.%20Namespaces%20and%20external%20types/Using%20attributes.cs#L181).

RunDevelopment commented:
Because the console object has a set of properties:

- Object.keys() will list all methods (log, debug, ...)
  aka. GetEnumerator will also enumerate these methods.
- console.prototype === undefined (aka its not a property of console)
- console.\_\_proto\_\_ === {} (empty JSObject)
- GetTypeName(console) === "Object"

In a nutshell: Behave like a JSObject where properties were added.
I'd also like to insert it directly into the GlobalContext. This will also eliminate the need for the static list of GlobalContexts.

RunDevelopment committed:
Fixed Arguments.GetEnumerator()

RunDevelopment committed:
Fixed LogArguments for args == null

RunDevelopment committed:
Fixed FormatArgs

RunDevelopment committed:
Fixed JSValueToObjectString and [Type].__proto__

RunDevelopment committed:
Array.prototype is now formatted correctly

nilproject commented:
`console.prototype` empty for all instances. This property is actual for constructors only.

As I can see, in Node.JS 6.9.5  `console.__proto__` not empty object.
```Console
Console {
  log: [Function],
  info: [Function],
  warn: [Function],
  error: [Function],
  dir: [Function],
  time: [Function],
  timeEnd: [Function],
  trace: [Function: trace],
  assert: [Function] }
```
So, I think, it's not prohibited to make some inconsistencies with the standard, if public part of API remains in accordance with it.

nilproject commented on NiL.JS/BaseLibrary/Console.cs:
```diff
@@ -47,22 +165,209 @@ public static void asserta(Function f, JSValue sample)
+
+        public static JSConsole CreateConsole(GlobalContext globalContext)
+        {
+            JSConsole c = GetConsole(globalContext);
```
Is it possible to create new instance of Console every time and not store it? As I can see, this methods calls only once for every GlobalContext.

RunDevelopment replied:
Yeah, sure, it will be. I want to get rid of the code handling the 1 JSConsole for 1 GlobalContext all together but I will wait until I found an elegant solution.
If possible I want to store the current JSConsole for a given GlobalContext, so that the user can access and modify it. This was the whole purpose of this structure. It's only a very unpleasant side effect that it disables you from creating new JSConsoles...

nilproject commented on NiL.JS/Core/Tools.cs:
```diff
@@ -1887,5 +1887,393 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+        }
+
+
+        public static string FormatArgs(IEnumerable<KeyValuePair<string, JSValue>> args)
```
This method does not use the keys, only values. `KeyValuePair<string, JSValue>` is redundant limitation. You can use `System.Linq`: `IEnumerable.Select(...)` on caller side.
I think this function is very specific. Let's make it internal. If this function will be public, then we must support it until the major version changes.

RunDevelopment replied:
Then maybe I should also rename this function because it doesn't really only formats arguments anymore.
How about this: It will try to convert every IEnumerable it gets into IEnumerable<JSValue>? I don't want to handle a case so common on the caller side. Plus, it will still be `FormatArgs([object Arguments])`, no conversion necessary.

RunDevelopment replied:
On word on new public functions:
All except for one of the functions I added to Tools are public; Do you want to make those internal as well? Some of them could be useful, e.g. in my little console, I use JSValueToObjectString with a high maxRecursionDepth to display output.
Also, functions like GetTypeName could be useful but are currently a little improvised. It works great for what it is used now but we might be able to break it.
Speaking of GetTypeName, maybe we should replace the code in JSValue.toString for objects with it to have the code in a central place.

RunDevelopment committed:
Updated FormatArgs

RunDevelopment committed:
Updated JSConsole to now be used in a wrapper

RunDevelopment committed:
Update for JSConsole

RunDevelopment commented:
Ok now we are left with (mainly) two problems:

### 1. The representation of console in JS

Currently, there is no way (at least no way I know of) to make an object that has these properties except for the method I used.
I also think that it is acceptable when _some_ properties are off but I want to keep it as close to the standard and the behaviour other engines like V8 or SpiderMonkey (please let's not count Chakra...).
To put it briefly, the main property out the four is the first one:

- Object.keys() will list all methods (log, debug, ...)
  aka. GetEnumerator will also enumerate these methods.

The other properties are basically extras.
I just really want that all properties/methods/... that are not marked as [Hidden] or [DoNotEnumerate] will be in Object.keys() aka that they get enumerated.

### 2. The one-to-one relationship between JSConsole and GlobalContext

This just has to go away...

### A possible solution

The easiest way is obviously just to pass it into JS, use an ObjectWrapper and make it (the console variable in JS) somehow overrideable so that the user can freely edit the console variable in JS and the JSConsole instance.

Problem: 
An ObjectWrapper doesn't enumerate everything I want and the ObjectWrapper, in general, feels a bit unfinished. We will probably extend the features of ObjectWrapper in the future to behave more javascript-like (at least I plan to) but for now, this might be the best way.

RunDevelopment commented:
What's left now is a public vs internal debate on Tools.

A note on that:
If FormatArgs is internal then the tests I added will be useless. This might be the only argument to keep it public but not a bad one because it will test the formatting on many edge cases which also led to the discovery of some bugs.

nilproject commented:
Tests are good. Just use reflection for it. It is normal for tests. Something like this:
```cs
var FormatArgs= (Func<IEnumerable, string>)typeof(Tools).GetMethod("FormatArgs", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic).CreateDelegate(typeof(Func<IEnumerable, string>));
```

RunDevelopment commented:
I will gladly do that. Thanks!

But what about the other functions? Internal as well?

RunDevelopment committed:
Updated FormatArgs test

RunDevelopment committed:
internal FormatArgs & JSValueToString

RunDevelopment committed:
Updated FormatArgs test

nilproject commented:
I think, `LogArguments` should be internal for the same reasons as `FormatArgs`

RunDevelopment committed:
Internal LogArguments

RunDevelopment commented:
Done.
Still, what about JSValueToObjectString() and GetTypeName() ?

nilproject commented:
These are useful functions. Let them be public.

RunDevelopment commented:
Ok. Thanks.
